File: routes/logistics_routes.py
"""
Routes cho dashboard hậu cần — Multi-tenant
"""
from flask import Blueprint, jsonify, request
from flask_login import current_user, login_required
from database import db, ThongBaoHauCan, GopY, DanhGia, ThucPhamThua, ChienSi, CatCom, HuongViQueNha, MonAn, get_user_don_vi_id
from datetime import date, datetime, timedelta
from sqlalchemy import func, extract

logistics_bp = Blueprint('logistics', __name__)


@logistics_bp.route('/api/logistics/dashboard')
@login_required
def get_dashboard():
    """Tổng hợp dashboard hậu cần (theo đơn vị)"""
    today = date.today()
    dvid = get_user_don_vi_id()
    
    # Thống kê đánh giá (theo đơn vị)
    rating_stats = db.session.query(
        func.avg(DanhGia.so_sao).label('tb'),
        func.count(DanhGia.id).label('tong')
    ).filter(DanhGia.don_vi_id == dvid).first()
    
    # Góp ý chưa đọc (theo đơn vị)
    gop_y_chua_doc = GopY.query.filter_by(da_doc=False, don_vi_id=dvid).count()
    
    # Thực phẩm thừa tuần này (theo đơn vị)
    start_week = today - timedelta(days=today.weekday())
    thua_tuan = db.session.query(
        func.sum(ThucPhamThua.luong_thua_kg)
    ).filter(ThucPhamThua.ngay >= start_week, ThucPhamThua.don_vi_id == dvid).scalar() or 0
    
    # Thực đơn hôm nay (theo đơn vị)
    from database import ThucDon
    so_bua_hom_nay = ThucDon.query.filter_by(ngay=today, don_vi_id=dvid).count()
    
    return jsonify({
        'danh_gia_trung_binh': round(float(rating_stats.tb or 0), 1),
        'tong_danh_gia': rating_stats.tong or 0,
        'gop_y_chua_doc': gop_y_chua_doc,
        'so_bua_hom_nay': so_bua_hom_nay,
        'thuc_pham_thua_tuan_kg': round(float(thua_tuan), 1)
    })


@logistics_bp.route('/api/logistics/announcements')
@login_required
def get_announcements():
    """Lấy danh sách thông báo hậu cần (theo đơn vị)"""
    dvid = get_user_don_vi_id()
    announcements = ThongBaoHauCan.query.filter_by(don_vi_id=dvid).order_by(
        ThongBaoHauCan.ghim.desc(),
        ThongBaoHauCan.ngay_tao.desc()
    ).all()
    return jsonify([a.to_dict() for a in announcements])


@logistics_bp.route('/api/logistics/announcements', methods=['POST'])
@login_required
def create_announcement():
    """Tạo thông báo hậu cần (gán đơn vị)"""
    dvid = get_user_don_vi_id()
    data = request.json
    tb = ThongBaoHauCan(
        tieu_de=data['tieu_de'],
        noi_dung=data['noi_dung'],
        don_vi_id=dvid,
        loai=data.get('loai', 'thong_bao'),
        gop_y_id=data.get('gop_y_id'),
        ghim=data.get('ghim', False)
    )
    db.session.add(tb)
    db.session.commit()
    return jsonify(tb.to_dict()), 201


@logistics_bp.route('/api/logistics/satisfaction-report')
@login_required
def get_satisfaction_report():
    """Báo cáo hài lòng theo tháng (theo đơn vị)"""
    dvid = get_user_don_vi_id()
    months = request.args.get('months', 3, type=int)
    
    result = []
    for i in range(months):
        d = date.today().replace(day=1) - timedelta(days=30 * i)
        month_start = d.replace(day=1)
        if d.month == 12:
            month_end = d.replace(year=d.year + 1, month=1, day=1) - timedelta(days=1)
        else:
            month_end = d.replace(month=d.month + 1, day=1) - timedelta(days=1)
        
        stats = db.session.query(
            func.avg(DanhGia.so_sao).label('tb'),
            func.count(DanhGia.id).label('tong'),
            func.sum(db.case((DanhGia.so_sao >= 4, 1), else_=0)).label('tot'),
            func.sum(db.case((DanhGia.so_sao <= 2, 1), else_=0)).label('kem')
        ).filter(
            DanhGia.don_vi_id == dvid,
            DanhGia.ngay_tao >= datetime.combine(month_start, datetime.min.time()),
            DanhGia.ngay_tao <= datetime.combine(month_end, datetime.max.time())
        ).first()
        
        result.append({
            'thang': f"{d.year}-{d.month:02d}",
            'trung_binh': round(float(stats.tb or 0), 1),
            'tong_danh_gia': stats.tong or 0,
            'danh_gia_tot': stats.tot or 0,
            'danh_gia_kem': stats.kem or 0
        })
    
    return jsonify(result)


@logistics_bp.route('/api/hometown-flavor')
@login_required
def get_hometown_flavor():
    """Lấy thông tin hương vị quê nhà tháng này (theo đơn vị)"""
    dvid = get_user_don_vi_id()
    today = date.today()
    hvqn = HuongViQueNha.query.filter_by(thang=today.month, nam=today.year, don_vi_id=dvid).first()
    
    # Chiến sĩ sinh nhật tháng này (trong đơn vị)
    birthday_soldiers = ChienSi.query.filter(
        extract('month', ChienSi.ngay_sinh) == today.month,
        ChienSi.don_vi_id.in_(
            db.session.query(db.distinct(ChienSi.don_vi_id)).filter(
                ChienSi.don_vi_id == dvid
            ) if dvid else []
        )
    ).all() if dvid else []
    
    # Phân bố vùng miền
    vung_mien_count = {}
    for cs in birthday_soldiers:
        vm = cs.vung_mien or 'chung'
        vung_mien_count[vm] = vung_mien_count.get(vm, 0) + 1
    
    return jsonify({
        'huong_vi': hvqn.to_dict() if hvqn else None,
        'chien_si_sinh_nhat': [cs.to_dict() for cs in birthday_soldiers],
        'so_sinh_nhat': len(birthday_soldiers),
        'phan_bo_vung_mien': vung_mien_count
    })

File: routes/logisticsRoutes.js
/**
 * Logistics routes - Dashboard hậu cần
 */
const express = require('express');
const router = express.Router();
const { db, getUserDonViId, getMonAnForThucDon } = require('../database');

const loginRequired = (req, res, next) => {
  if (!req.session?.user) return res.status(401).json({ error: 'Chua dang nhap' });
  next();
};

router.get('/api/logistics/dashboard', loginRequired, async (req, res) => {
  try {
    const today = new Date().toISOString().split('T')[0];
    const dvid = getUserDonViId(req);
    const ratingStats = await db.get('SELECT AVG(so_sao) as tb, COUNT(id) as tong FROM danh_gia WHERE don_vi_id = ?', dvid);
    const gopYChuaDoc = (await db.get('SELECT COUNT(*) as c FROM gop_y WHERE da_doc = 0 AND don_vi_id = ?', dvid)).c;
    const d = new Date(); d.setDate(d.getDate() - d.getDay() + (d.getDay() === 0 ? -6 : 1));
    const startWeek = d.toISOString().split('T')[0];
    const thuaTuan = (await db.get('SELECT COALESCE(SUM(luong_thua_kg), 0) as s FROM thuc_pham_thua WHERE ngay >= ? AND don_vi_id = ?', startWeek, dvid)).s;
    const soBuaHomNay = (await db.get('SELECT COUNT(*) as c FROM thuc_don WHERE ngay = ? AND don_vi_id = ?', today, dvid)).c;
    res.json({
      danh_gia_trung_binh: Math.round((parseFloat(ratingStats.tb) || 0) * 10) / 10,
      tong_danh_gia: parseInt(ratingStats.tong) || 0,
      gop_y_chua_doc: parseInt(gopYChuaDoc) || 0,
      so_bua_hom_nay: parseInt(soBuaHomNay) || 0,
      thuc_pham_thua_tuan_kg: Math.round((parseFloat(thuaTuan) || 0) * 10) / 10
    });
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.get('/api/logistics/announcements', loginRequired, async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    const list = await db.all('SELECT * FROM thong_bao_hau_can WHERE don_vi_id = ? ORDER BY ghim DESC, ngay_tao DESC', dvid);
    res.json(list.map(a => ({
      id: a.id, tieu_de: a.tieu_de, noi_dung: a.noi_dung, loai: a.loai,
      gop_y_id: a.gop_y_id, ghim: !!a.ghim, ngay_tao: a.ngay_tao
    })));
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.post('/api/logistics/announcements', loginRequired, async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    const { tieu_de, noi_dung, loai = 'thong_bao', gop_y_id, ghim = false } = req.body;
    const r = await db.run('INSERT INTO thong_bao_hau_can (tieu_de, noi_dung, don_vi_id, loai, gop_y_id, ghim) VALUES (?, ?, ?, ?, ?, ?)',
      tieu_de, noi_dung, dvid, loai, gop_y_id || null, ghim ? 1 : 0);
    const tb = await db.get('SELECT * FROM thong_bao_hau_can WHERE id = ?', r.lastInsertRowid);
    res.status(201).json({ id: tb.id, tieu_de: tb.tieu_de, noi_dung: tb.noi_dung, loai: tb.loai, gop_y_id: tb.gop_y_id, ghim: !!tb.ghim, ngay_tao: tb.ngay_tao });
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.get('/api/logistics/satisfaction-report', loginRequired, async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    const months = parseInt(req.query.months) || 3;
    const result = [];
    for (let i = 0; i < months; i++) {
      const d = new Date(); d.setMonth(d.getMonth() - i); d.setDate(1);
      const year = d.getFullYear(); const month = d.getMonth() + 1;
      const monthStart = `${year}-${String(month).padStart(2, '0')}-01`;
      const nextM = new Date(year, month, 1);
      const monthEnd = new Date(nextM - 86400000).toISOString().split('T')[0];
      const stats = await db.get(`SELECT AVG(so_sao) as tb, COUNT(id) as tong,
        SUM(CASE WHEN so_sao >= 4 THEN 1 ELSE 0 END) as tot,
        SUM(CASE WHEN so_sao <= 2 THEN 1 ELSE 0 END) as kem
        FROM danh_gia WHERE don_vi_id = ? AND ngay_tao >= ? AND ngay_tao <= ?`, dvid, monthStart, monthEnd + ' 23:59:59');
      result.push({
        thang: `${year}-${String(month).padStart(2, '0')}`,
        trung_binh: Math.round((parseFloat(stats.tb) || 0) * 10) / 10,
        tong_danh_gia: parseInt(stats.tong) || 0,
        danh_gia_tot: parseInt(stats.tot) || 0,
        danh_gia_kem: parseInt(stats.kem) || 0
      });
    }
    res.json(result);
  } catch (e) { res.status(500).json({ error: e.message }); }
});

router.get('/api/hometown-flavor', loginRequired, async (req, res) => {
  try {
    const dvid = getUserDonViId(req);
    const today = new Date();
    const month = today.getMonth() + 1;
    const year = today.getFullYear();
    const hvqn = await db.get('SELECT * FROM huong_vi_que_nha WHERE thang = ? AND nam = ? AND don_vi_id = ?', month, year, dvid);
    let hvqnDict = null;
    if (hvqn) {
      const ma = await db.get('SELECT * FROM mon_an WHERE id = ?', hvqn.mon_an_id);
      hvqnDict = { id: hvqn.id, mon_an_id: hvqn.mon_an_id, mon_an: ma || null, thang: hvqn.thang, nam: hvqn.nam, mo_ta: hvqn.mo_ta };
    }
    const IS_PG = !!process.env.DATABASE_URL;
    let birthdaySoldiers = [];
    if (dvid) {
      if (IS_PG) {
        birthdaySoldiers = await db.all(
          `SELECT * FROM chien_si WHERE EXTRACT(MONTH FROM ngay_sinh::date)::integer = ? AND don_vi_id = ?`, month, dvid);
      } else {
        birthdaySoldiers = await db.all(
          `SELECT * FROM chien_si WHERE CAST(strftime('%m', ngay_sinh) AS INTEGER) = ? AND don_vi_id = ?`, month, dvid);
      }
    }
    const vungMienCount = {};
    birthdaySoldiers.forEach(cs => {
      const vm = cs.vung_mien || 'chung';
      vungMienCount[vm] = (vungMienCount[vm] || 0) + 1;
    });
    res.json({
      huong_vi: hvqnDict,
      chien_si_sinh_nhat: birthdaySoldiers.map(cs => ({
        id: cs.id, ho_ten: cs.ho_ten, cap_bac: cs.cap_bac, chuc_vu: cs.chuc_vu,
        que_quan: cs.que_quan, vung_mien: cs.vung_mien, ngay_sinh: cs.ngay_sinh,
        don_vi_id: cs.don_vi_id, trang_thai: cs.trang_thai
      })),
      so_sinh_nhat: birthdaySoldiers.length,
      phan_bo_vung_mien: vungMienCount
    });
  } catch (e) { res.status(500).json({ error: e.message }); }
});

module.exports = router;
